Blog post part 1 about MyST and Nikola (#4)
damianavila/damian_blog: posts/draft.md

# A deep dive into MyST, Part 1: The MyST-Parser Python API usage in Nikola

> > *This is a series of blog posts inviting you to join me in a little journey I have
> experienced in the last few weeks at the time to figure it out a nice story for **MyST**
> inside the **Nikola** world.*

Reviewer: ```suggestion
> *This is a series of blog posts inviting you to join me in a little journey I have
experienced in the last few weeks to figure out a nice story for **MyST**
inside the **Nikola** world.*
```

Hey, a lot of time without writing, you know... (lack of) time, but let's go back to the
important things... or at least, some interesting stuff to share with you!

Recently, I have been involved with the [Executable Books](https://executablebooks.org/en/latest/)
project and community (you can learn more about them [here](https://executablebooks.org/en/latest/about.html)).
One of the major projects inside that community is the [MyST-Parser](https://github.com/executablebooks/MyST-Parser),
a *rich* and *extensible* flavor of Markdown meant for technical documentation and
publishing. It supports all the syntax of CommonMark Markdown at its base but, in
addition, it includes several extensions to CommonMark adding extra syntax features for
technical writing, such as the roles and directives used by Sphinx in the rST world!

The whole Executable Books project is about providing an excellent pipeline for

> publishing Jupyter Notebook, with the MyST-Parser being one of the central building

Reviewer: ```suggestion
publishing Jupyter Notebooks, with the MyST-Parser being one of the central building
```

have support for the main pieces they use on a daily basis to save them some time.

So, what happens when you try to use the MyST Python API to render a simple directive:

```ipython
In [1]: from myst_parser.main import to_html

In [2]: text = """
   ...: ```{admonition} This is my admonition
   ...: This is my note
   ...: ```
   ...: """

In [3]: to_html(text)
Out[3]: '<pre><code class="language-{admonition}">This is my note\n</code></pre>\n'
```

That HTML code does not look like an admonition!
When I searched into the MyST and Nikola issue trackers, I found a [report](https://github.com/executablebooks/MyST-Parser/issues/307)
about this problem.

One of the answers from a MyST core contributor was particularly interesting:

> > I don't know if it's intended to behave this way.

> Heya, yes it is; you have to use sphinx to fully parse to HTML

That surprised me, as I assumed that roles and directives would be supported by the
Python API, but I learnt things are a little bit more complex than that!

## Understanding the limitations (fun/not fun dive, actually fun...)

I started digging into the codebase and some MyST documentation.

> If you look at the `to_html` function I shown before: 

Reviewer: ```suggestion
If you look at the `to_html` function I showed before: 
```

If you indeed keep looking at the `default_parser` implementation, you will find the
MyST-Parser is actually a collection of `markdown-it-py` plugins:

```python
    md = (
        MarkdownIt("commonmark", renderer_cls=renderer_cls)
        .enable("table")
        .use(front_matter_plugin)
        .use(myst_block_plugin)
        .use(myst_role_plugin)
        .use(footnote_plugin)
        .use(wordcount_plugin, per_minute=config.words_per_minute)
        .disable("footnote_inline")
        .disable("footnote_tail")
    )
```

Notice there are some MyST specific plugins such as the `myst_block_plugin` and the
`myst_role_plugin`.

So, it seems the main idea is to have a `MystParser`, a Sphinx parser, that parses stuff
into a token stream using a set of general and some specific markdown-it-py plugins,
creates a Docutils objects-based representation of the token stream and render those
objects using the Sphinx machinery.

In this model, it actually makes sense to see roles and directives supported when you go
through the Docutils/Sphinx path (those concepts are core-supported in the
Docutils/Sphinx ecosystem) and those unsupported when you use the MyST Python API which

> does not understand how to parse nor render these concepts as we previously shown.

Reviewer: ```suggestion
does not understand how to parse nor render these concepts as we've previously shown.
```

> # A deep dive into the MyST world

Reviewer: ```suggestion
# A deep dive into MyST Markdown, Part 1: Roles and directives in Nikola
```

Reviewer: I think this would make the title a bit more specific so that you could build upon it for your other parts. (feel free to re-work the wording if you like)

Author: > (feel free to re-work the wording if you like)

OK, let me think about this one...

> > *This is a series of blog posts inviting you to join me in a little journey I have experienced in the last few weeks at the time to figure it out a nice story for **MyST**

Reviewer: Just a meta thought - I like using the markdown approach of _linebreak after every sentence_ so that you don't run into semi-arbitrary line breaks like this

Author: Mmm... I need to try it... I see how linebreak after every sentence could be nice for some people and really ugly for others. I promise I will eventually make a try with another piece of text :wink:

Reviewer: in case it's helpful, I've found this site to be a useful rationale for doing line breaks this way :-) https://sembr.org/

they actually recommend more complex rules for deciding when to break lines, but I think that "one line per sentence" is easy enough to follow

inside the **Nikola** world.*

Hey, a lot of time without writing, you know... (lack of) time, but let's go back to the
important things... or at least, interesting stuff to share with you!

> Recently, I have been involved with the [EBP](https://executablebooks.org/en/latest/)

Reviewer: ```suggestion
Recently, I have been involved with the [Executable Books](https://executablebooks.org/en/latest/)
```

project and community (you can learn more about them [here](https://executablebooks.org/en/latest/about.html)). One of the major projects inside that community is the
[MyST-Parser](https://github.com/executablebooks/MyST-Parser), a *rich* and *extensible*

> flavor of Markdown meant for technical documentation and publishing. It supports all thesyntax of CommonMark Markdown at its base but, in addition, it includes several

Reviewer: ```suggestion
flavor of Markdown meant for technical documentation and publishing. It supports all the syntax of CommonMark Markdown at its base but, in addition, it includes several
```

Reviewer: there are a few other places where you've forgotten a space, I'll try to find them but you should double-check

Author: OK, I will pay attention to other occurrences!

extensions to CommonMark adding extra syntax features for technical writing, such as the
roles and directives used by Sphinx in the rST world!

> When I looked into this project, I quickly realized it would interesting to explore how

Reviewer: I'd try and reduce the number of adverbs that you use. I know that you are shooting for a "familiar" tone here, and in general they are OK, but using them so much makes it a bit harder to follow the flow of sentences.

```suggestion
When I looked into this project, I realized it would interesting to explore how
```

this flavor of Markdown could potentially integrate with the [Nikola](https://getnikola.com/)
ecosystem. If you look into my previous blog posts, you will see I am a Nikola
user and contributor and I am always interested in how we can improve the

> "blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circle

Reviewer: A meta-question: did you also look into using Jupytext for blogging with MyST Jupyter Notebooks in Nikola? This could be a nice way to keep the structure of notebooks, but with the benefit of version control

Author: I briefly look into it but did not go full on that because it was not giving enough exposure to the MyST machinery, IMHO.
I think it would be a really nice topic for another (and related) blog post, actually.

> "blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circle
> around an excel pipeline for publishing Jupyter Notebook being the Myst-Parser one of
> the central building blocks. They are even [interested](https://github.com/executablebooks/jupyter-book/issues/900)

Reviewer: ```suggestion
"blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circles
around an excellent pipeline for publishing Jupyter Notebook, with the MyST-Parser being one of
the central building blocks. They are even [interested](https://github.com/executablebooks/jupyter-book/issues/900)
```

in a sane blog story for one of the key pieces the community has, the
[Jupyter Book](https://jupyterbook.org/intro.html) project that indeed uses the
Myst-Parser!

Nikola, at the same time, has a long story providing support for
multiple flavors of Markdown as Nikola plugins. Additionally, Nikola is probably unique
in the Python static site generators arena because it is supporting Jupyter Notebook
from the core package!

So, all the pieces were asking me to explore this space and I could not resist ;-)

# First impressions (someone has beaten me to it...)

> Ok, the first idea coming to my mind was to somehow implement MyST support in Nikola.

Reviewer: Are you OK with me making small english changes like this? 
```suggestion
The first idea that came to my mind was to somehow implement MyST support in Nikola.
```

But, hey, somebody probably already thought about this before me. So I made a search to
see if this was not previously explored by others. And indeed that was the case!

> Roberto, the Nikola creator, wrote a plugin for MyST about a year ago (there is a brief
> blog post talking about it [here](https://ralsina.me/tr/es/weblog/posts/new-plugins-for-nikola.html)).

Reviewer: Try to hyperlink more of the text so that it is easier for people to notice. For example, [the sentence here](https://executablebooks.org) is easier to notice, whereas this sentence [here](https://executablebooks.org) is harder to notice.
```suggestion
Roberto, the Nikola creator, wrote a plugin for MyST about a year ago (there is [a brief
blog post talking about it](https://ralsina.me/tr/es/weblog/posts/new-plugins-for-nikola.html)).
```

That's great, isn't it? Obviously, I quickly jumped into looking at the codebase for that
plugin and the core of the implementation was quite simple actually:

```python
output = myst_parser.main.to_html(new_data)
```

As you can see, it is just a matter to use the Python MyST API to generate the HTML
output the Nikola machinery is going to write to build the blog post.

Problem solved! Someone else had the idea, implemented it and now I can use it, right?
Well... actually it was not that simple...

> # Current limitations (there is always a but...)

Reviewer: This should be a sub-section, right? 
```suggestion
## Current limitations (there is always a but...)
```

tempted to use MyST and the MyST plugin for Nikola if they have support for the main
pieces they use on a daily basis to save some time, right?

So what happens when you try to use the MyST Python API to render a simple directive:

```ipython
In [1]: from myst_parser.main import to_html

In [2]: text = """
   ...: ```{admonition} This is my admonition
   ...: This is my note
   ...: ```
   ...: """

In [3]: to_html(text)
Out[3]: '<pre><code class="language-{admonition}">This is my note\n</code></pre>\n'
```

Mmm... that HTML code does not look like an admonition!
When I quickly search into the MyST and Nikola issue trackers, I quickly found some
[report](https://github.com/executablebooks/MyST-Parser/issues/307) about this problem.

One of the answers from a core contributor was particularly interesting...

> > I don't know if it's intended to behave this way.

> Heya, yes it is; you have to use sphinx to fully parse to HTML

Wait? Why? I was expecting the MyST Python API to produce the expected HTML for roles
and directives and that is not supported? Is the API incomplete? Is this intended by
design? A lot of questions came to my mind and, as usual, the answer was in the

> documentation and the codebase.

Reviewer: Just a note that this might come across as kind-of negative on the myst parser...it might be better to just get straight to the point, e.g. "This surprised me, as I assumed that roles and directives would be supported by the Python API, but I learned that it is more complicated than that."

Author: Thank you... the negativity you mentioned did not pass through my head at any time. I will do the change as requested.

later). So, let get into those parser internals:

```python
def default_parser(config: MdParserConfig) -> MarkdownIt:
    """Return the default parser configuration for MyST"""
    renderer_cls: Callable[[MarkdownIt], RendererProtocol]

    if config.renderer == "sphinx":
        from myst_parser.sphinx_renderer import SphinxRenderer

        renderer_cls = SphinxRenderer
    elif config.renderer == "html":
        renderer_cls = RendererHTML
    elif config.renderer == "docutils":
        from myst_parser.docutils_renderer import DocutilsRenderer

        renderer_cls = DocutilsRenderer
    else:
        raise ValueError("unknown renderer type: {0}".format(config.renderer))
```

Well, this is quite interesting! We have a `SphinxRenderer` and `DocutilsRenderer`. You
can learn about those [here](https://myst-parser.readthedocs.io/en/latest/api/renderers.html#).
Those are an integral part of the Myst-Parser project. But when you look at the `html`
renderer, that [RendererHTML] class (https://markdown-it-py.readthedocs.io/en/latest/api/markdown_it.renderer.html#markdown_it.renderer.RendererHTML)
is coming from another EBP project called markdown-it-py. This is a Python port of the

> well-used [markdown-it](https://github.com/markdown-it/markdown-it), plus some of itsassociated plugins.

Reviewer: ```suggestion
well-used [markdown-it](https://github.com/markdown-it/markdown-it), plus some of its associated plugins.
```

If you indeed keep looking at the `default_parser` implementation, you will find the
MyST-Parser is actually a collection of markdown-it-py plugins:

```python
    md = (
        MarkdownIt("commonmark", renderer_cls=renderer_cls)
        .enable("table")
        .use(front_matter_plugin)
        .use(myst_block_plugin)
        .use(myst_role_plugin)
        .use(footnote_plugin)
        .use(wordcount_plugin, per_minute=config.words_per_minute)
        .disable("footnote_inline")

>         # disable this for now, because it need a new implementation in the renderer

Reviewer: Just to avoid readers needing to wonder what this comment means
```suggestion
```

        .disable("footnote_tail")
    )
```

Notice there are some MyST specific plugins such as the `myst_block_plugin` and the
`myst_role_plugin`. The implementation is more complex, of course, I am just extracting
the core idea ;-)

So, it seems the main idea is to have a `MystParser`, a Sphinx parser, that parses stuff
into a token stream using a set of general and some specific markdown-it-py plugins,
creates a Docutils objects-based representation of the token stream and render those
objects using the Sphinx machinery.

In this model, it actually makes sense to see roles and directives supported when you go
through the Docutils/Sphinx path (those concepts are core-supported in the
Docutils/Sphinx ecosystem) and those unsupported when you use the MyST Python API which
does not understand how to parse nor render these concepts as we saw previously.

# Overcome the limitations (don't worry I will stop here... for now)

OK, this is long enough for now, the next post will deep dive a little bit more in the

> Myst-parser and the markdown-it-py codebases in the context of potentially supporting

Reviewer: ```suggestion
MyST-parser and the markdown-it-py codebases in the context of potentially supporting
```

Reviewer: I think it should be either `myst` (all lowercase) or `MyST`, but not intermixed (`Myst`)

Author: I totally agree!
